src/persist/src/client.rs

```diff
@@ -393,85 +396,80 @@ impl<'a, K: Codec, V: Codec> FromIterator<&'a ((K, V), u64, isize)> for WriteReq
-            handle.client.clone()
-        } else {
-            return Err(Error::from("MultiWriteHandle received no streams"));
-        };
-        for handle in handles.iter() {
-            if handle.client != client {
-                return Err(Error::from(format!(
-                    "MultiWriteHandle got handles from two runtimes: {:?} and {:?}",
-                    client, handle.client
-                )));
-            }
-            // It's odd if there are duplicates but the semantics of what that
-            // means are straightforward, so for now we support it.
-            stream_ids.insert(handle.stream_id()?);
+    ///
+    /// Convenience function for when all the streams have matching K and V.
+    pub fn new_from_streams<K, V, H, I>(mut handles: I) -> Result<Self, Error>
+    where
+        K: Codec,
+        V: Codec,
+        H: Borrow<StreamWriteHandle<K, V>>,
+        I: Iterator<Item = H>,
+    {
+        // Manually pop the first handle to construct the MultiWriteHandle, then
+        // continue with the same iter to add the rest.
+        let first_handle = handles
+            .next()
+            .ok_or_else(|| Error::from("MultiWriteHandle requires at least one stream"))?;
+        let mut ret = Self::new(first_handle.borrow())?;
+        for handle in handles {
+            ret.add_stream(handle.borrow())?;
         }
-        Ok(MultiWriteHandle {
-            stream_ids,
-            client,
-            _phantom: PhantomData,
-        })
+        Ok(ret)
+    }
+
+    /// Adds the given stream to the set
+    pub fn add_stream<K: Codec, V: Codec>(
+        &mut self,
+        handle: &StreamWriteHandle<K, V>,
+    ) -> Result<(), Error> {
+        if handle.client != self.client {
+            return Err(Error::from(format!(
+                "MultiWriteHandle got handles from two runtimes: {:?} and {:?}",
+                self.client, handle.client
+            )));
+        }
+        // It's odd if there are duplicates but the semantics of what that
+        // means are straightforward, so for now we support it.
+        self.stream_ids.insert(handle.stream_id()?);
+        Ok(())
     }
 
     /// Atomically writes the given updates to the paired streams.
     ///
     /// Either all of the writes will be made durable for replay or none of them
     /// will.
-    ///
-    /// Ids may be duplicated. However, the updates are passed down to storage
-    /// unchanged, so users should coalesce them when that's not otherwise
-    /// slower.
-    //
-    // TODO: This could take &StreamWriteHandle instead of Id to avoid surfacing
-    // Id to users, but that would require an extra Vec. Revisit.
-    //
-    // TODO: Make this take a two-layer IntoIterator to mirror how
-    // StreamWriteHandle::write works.
-    pub fn write_atomic(&self, updates: Vec<(Id, Vec<((K, V), u64, isize)>)>) -> PFuture<SeqNo> {
+    pub fn write_atomic<F: FnOnce(&mut AtomicWriteBuilder<'_>) -> Result<(), Error>>(
```

Comment:
Why did you go for the builder instead of a two-layer `IntoIterator` as mentioned in the `TODO`?

I'm just curious, plus I don't mind the builder pattern.

Reply:
I still would prefer the two-level IntoIterator but it doesn't work when the (K, V) bounds don't match, which is something I wanted to support

Reply:
:+1: